### PR TITLE
feat(ses): call lockdown before bundling SES shim

### PR DESCRIPTION
**Pull request by leotm**
## Description

Call lockdown before bundling the SES shim

Discussion: https://youtu.be/qGGeACz4cyM

### Security Considerations

> Does this change introduce new assumptions or dependencies that, if violated, could introduce security vulnerabilities? How does this PR change the boundaries between mutually-suspicious components? What new authorities are introduced by this change, perhaps by new API calls?

- bundle now fails (TypeError) to build if non-standard JS called after `lockdown()`
- `yarn build` now also logs to console (e.g. `Removing unpermitted intrinsics`) if non-standard JS called before `lockdown()`

### Scaling Considerations

> Does this change require or encourage significant increase in consumption of CPU cycles, RAM, on-chain storage, message exchanges, or other scarce resources? If so, can that be prevented or mitigated?

### Documentation Considerations

> Give our docs folks some hints about what needs to be described to downstream users.  Backwards compatibility: what happens to existing data or deployments when this code is shipped? Do we need to instruct users to do something to upgrade their saved data? If there is no upgrade path possible, how bad will that be for users?

### Testing Considerations

> Every PR should of course come with tests of its own functionality. What additional tests are still needed beyond those unit tests? How does this affect CI, other test automation, or the testnet? 

### Compatibility Considerations

> Does this change break any prior usage patterns? Does this change allow usage patterns to evolve?

### Upgrade Considerations

> What aspects of this PR are relevant to upgrading live production systems, and how should they be addressed?

> Include `*BREAKING*:` in the commit message with migration instructions for any breaking change.

> Update `NEWS.md` for user-facing changes.

> Delete guidance from pull request description before merge (including this!)

### CODE REVIEW

File: packages/ses/scripts/bundle.js
Comment (erights) [inline packages/ses/scripts/bundle.js:4]: IIUC, this does not actually change the order in which things happen, at least in the absence of cyclic imports. All imported modules are evaluated/initialized before the code in this module starts evaluation/initialization. So either way, `../index.js` initialization will happen before this call to `lockdown`.

Attn @kriskowal

**Review by erights — CHANGES_REQUESTED** [review]

File: packages/ses/scripts/bundle.js
Comment (kriskowal) [reply packages/ses/scripts/bundle.js:4]: This is true. To interleave behavior relative to other imports, we have to use a module like `lockdown.js` that just calls `lockdown()`. So,

```js
import '../index.js'; // or maybe import 'ses' reflexively works now
import './lockdown.js'; // just lockdown()
```

This insures that `./index.js` initializes before you call `lockdown()`.

It’s unclear why the tests pass with this defect. It might be that we need to test `yarn build` under `yarn test`.

Comment (leotm) [reply packages/ses/scripts/bundle.js]: makes sense ^ reverted [feat(ses): call lockdown earlier](https://github.com/endojs/endo/pull/2337/commits/b5c04714ebdfa5a36598b0b21f9c3d31d0525c73), looks good now

File: packages/ses/scripts/bundle.js
Comment (leotm) [inline packages/ses/scripts/bundle.js:4]:
```suggestion
// eslint-disable-next-line import/no-extraneous-dependencies
import 'ses';
```

also seems ok here

File: packages/ses/scripts/bundle.js
Comment (leotm) [reply packages/ses/scripts/bundle.js]: > It’s unclear why the tests pass with this defect. It might be that we need to test `yarn build` under `yarn test`.

yeah the tests still pass calling lockdown() too early before the imports 🤔

https://github.com/endojs/endo/blob/7a2955080f34880c8028e1b2dff06bf7b969790f/packages/ses/package.json#L75

even when doing a clean build before `"test": "yarn prepare && tsd && ava",` if that's what we meant

**Review by erights — APPROVED** [review]
I'm approving to clear my previous "request changes". FWIW LGTM, but on this I defer to @kriskowal . Please wait for an approval from him.

File: packages/ses/scripts/bundle.js
Comment (kriskowal) [inline packages/ses/scripts/bundle.js:8]: This has the effect of ensuring that `lockdown()` gets called before the following imports evaluate:
```suggestion
import '../test/lockdown-safe.js';
```

The proposed change causes `lockdown()` to be called after all imported modules have evaluated. This is one of surprising effects of ESM loader semantics. Imported modules do not evaluate on the stack of the `import` statement in the way that CJS modules _may_ evaluate on the stack of a `require` call. CJS bundlers generally behave like ESM, though.

Even with the above formulation, it is possible to force the subsequent imports to evaluate earlier (and thus in the pre-lockdown realm) by importing this module from another module with a different import order over the same modules. This is outside our concern because this is the entrypoint module for the program.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Posted the required change. Thank you for engaging!

**Review by kriskowal — APPROVED** [review]

**Pull request merged**
